File: services/email_service.py
"""
Production-ready Email Service for Campus Cognition V2.
Supports SMTP and API-key based providers.
Sends multi-part HTML and plain-text emails.
"""
import os
import smtplib
from email.message import EmailMessage
import logging
from flask import render_template, current_app

logger = logging.getLogger(__name__)

def _get_email_config():
    """Retrieve and normalize email configuration from environment variables."""
    host = os.getenv('MAIL_HOST', '').strip()
    port = int(os.getenv('MAIL_PORT', '587'))
    username = os.getenv('MAIL_USERNAME', '').strip()
    password = os.getenv('MAIL_PASSWORD', '')
    api_key = os.getenv('MAIL_API_KEY', '').strip()
    
    # Use API key as password for API-based SMTP providers (e.g., SendGrid, Mailgun)
    if api_key and not password:
        password = api_key
        if not username:
            username = 'apikey' # Default for many providers
            
    from_email = os.getenv('MAIL_FROM', '').strip()
    from_name = os.getenv('MAIL_FROM_NAME', 'Campus Cognition').strip()
    use_tls = os.getenv('MAIL_USE_TLS', 'true').lower() == 'true'
    
    sender = f"{from_name} <{from_email}>" if from_name else from_email
    
    return {
        'host': host,
        'port': port,
        'username': username,
        'password': password,
        'sender': sender,
        'from_email': from_email,
        'use_tls': use_tls
    }

def send_email(recipient: str, subject: str, body_text: str, body_html: str = None) -> bool:
    """Send an email using configured provider with safe failure."""
    config = _get_email_config()
    
    if not config['host'] or not config['from_email']:
        logger.warning('Email not sent: MAIL_HOST and MAIL_FROM are not configured.')
        return False
        
    message = EmailMessage()
    message['Subject'] = subject
    message['From'] = config['sender']
    message['To'] = recipient
    
    # Set plaintext body
    message.set_content(body_text)
    
    # Add HTML alternative if provided
    if body_html:
        message.add_alternative(body_html, subtype='html')
        
    try:
        # A mail provider outage must not leave the auth pages loading indefinitely
        # Use a short timeout to fail fast.
        with smtplib.SMTP(config['host'], config['port'], timeout=10) as server:
            if config['use_tls']:
                server.starttls()
            if config['username'] and config['password']:
                server.login(config['username'], config['password'])
            server.send_message(message)
        logger.info(f"Email sent successfully to {recipient}")
        return True
    except (OSError, smtplib.SMTPException) as error:
        logger.error(f"Email delivery failed: {error.__class__.__name__}")
        # Never log actual credentials or the exact recipient in full detail for security
        return False

def send_welcome_email(user: dict) -> bool:
    """Send welcome email to a newly registered user."""
    name = user.get('first_name') or user.get('username')
    email = user.get('email')
    
    subject = "Welcome to Campus Cognition 🎓"
    
    # Text fallback
    body_text = f"""Hello {name},

Welcome to Campus Cognition.
Your account has been successfully created.

Registered email: {email}

You can now use Campus Cognition to:
• Analyze your syllabus and PYQs
• Build personalized study roadmaps
• Ask questions about your notes
• Review and optimize code
• Discover internships and scholarships
• Track career opportunities
• Use Campus AI

For security, your password is never stored or sent by email.

Regards,
Campus Cognition Team
"""
    
    body_html = None
    try:
        if current_app:
            body_html = render_template('emails/welcome.html', name=name, email=email)
    except Exception as e:
        logger.warning(f"Could not render HTML template for welcome email: {e}")
        
    return send_email(email, subject, body_text, body_html)

def send_password_reset_email(user: dict, reset_url: str) -> bool:
    """Send a secure password reset link."""
    name = user.get('first_name') or user.get('username')
    email = user.get('email')
    
    subject = "Reset your Campus Cognition password"
    
    body_text = f"""Hello {name},

We received a request to reset your Campus Cognition password.

Use this one-time link to create a new password:
{reset_url}

This link expires in 30 minutes and can only be used once.

If you did not request this reset, you can safely ignore this email.
For security, Campus Cognition never sends your existing password by email.

Regards,
Campus Cognition Team
"""
    
    body_html = None
    try:
        if current_app:
            body_html = render_template('emails/password_reset.html', name=name, reset_url=reset_url)
    except Exception as e:
        logger.warning(f"Could not render HTML template for password reset email: {e}")

    return send_email(email, subject, body_text, body_html)

def send_password_changed_email(user: dict) -> bool:
    """Notify user that their password was changed."""
    name = user.get('first_name') or user.get('username')
    email = user.get('email')
    
    subject = "Your Campus Cognition password was changed"
    
    body_text = f"""Hello {name},

Your Campus Cognition password was successfully changed.

If you did not make this change, please contact support immediately.

Regards,
Campus Cognition Team
"""
    
    body_html = None
    try:
        if current_app:
            body_html = render_template('emails/password_changed.html', name=name)
    except Exception as e:
        logger.warning(f"Could not render HTML template for password changed email: {e}")

    return send_email(email, subject, body_text, body_html)

def send_application_deadline_email(user: dict, opportunity: dict) -> bool:
    """Notify user of an upcoming application deadline."""
    name = user.get('first_name') or user.get('username')
    email = user.get('email')
    
    subject = f"Deadline Approaching: {opportunity.get('title')}"
    
    body_text = f"""Hello {name},

This is a reminder that the application deadline for '{opportunity.get('title')}' at {opportunity.get('company', 'Organization')} is approaching on {opportunity.get('deadline')}.

Don't forget to submit your application!

Regards,
Campus Cognition Team
"""
    
    body_html = None
    try:
        if current_app:
            body_html = render_template('emails/application_deadline.html', name=name, opportunity=opportunity)
    except Exception as e:
        logger.warning(f"Could not render HTML template for deadline email: {e}")

    return send_email(email, subject, body_text, body_html)

def send_notification_email(user: dict, subject: str, message: str) -> bool:
    """Send a generic notification email."""
    name = user.get('first_name') or user.get('username')
    email = user.get('email')
    
    body_text = f"""Hello {name},

{message}

Regards,
Campus Cognition Team
"""
    
    body_html = None
    try:
        if current_app:
            body_html = render_template('emails/notification.html', name=name, message=message)
    except Exception as e:
        logger.warning(f"Could not render HTML template for notification email: {e}")

    return send_email(email, subject, body_text, body_html)

File: server/services/emailService.js
const nodemailer = require('nodemailer');
const env = require('../config/env');

let transporter = null;

if (env.mailUsername && env.mailPassword) {
  transporter = nodemailer.createTransport({
    host: env.mailHost,
    port: env.mailPort,
    secure: env.mailPort === 465,
    auth: {
      user: env.mailUsername,
      pass: env.mailPassword
    },
    tls: {
      rejectUnauthorized: false
    }
  });
}

async function sendWelcomeEmail(user) {
  if (!transporter) {
    console.log('[EmailService] SMTP credentials not set, skipping welcome email send.');
    return false;
  }

  try {
    const mailOptions = {
      from: `"${env.mailFrom || 'Campus Cognition'}" <${env.mailUsername}>`,
      to: user.email,
      subject: 'Welcome to Campus Cognition!',
      html: `
        <div style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; max-width: 600px; margin: 0 auto; padding: 24px; background-color: #161618; color: #ffffff; border-radius: 8px;">
          <h1 style="color: #ffffff; font-size: 24px; font-weight: bold; margin-bottom: 20px;">Welcome to Campus Cognition 🎓</h1>
          
          <p style="font-size: 16px; margin-bottom: 16px; color: #e5e5e5;">Hello ${user.firstName || user.username},</p>
          
          <p style="font-size: 16px; margin-bottom: 24px; line-height: 1.5; color: #e5e5e5;">
            Your account has been successfully created.<br>
            We're excited to have you on board!
          </p>

          <p style="font-size: 16px; margin-bottom: 24px; color: #e5e5e5;">
            <strong>Registered email:</strong><br>
            <a href="mailto:${user.email}" style="color: #60A5FA; text-decoration: none;">${user.email}</a>
          </p>

          <p style="font-size: 16px; margin-bottom: 16px; color: #e5e5e5;">You can now use Campus Cognition to:</p>
          
          <ul style="font-size: 15px; margin-bottom: 32px; padding-left: 20px; line-height: 1.8; color: #e5e5e5;">
            <li>Analyze your syllabus and PYQs</li>
            <li>Build personalized study roadmaps</li>
            <li>Ask questions about your notes</li>
            <li>Review and optimize code</li>
            <li>Discover internships and scholarships</li>
            <li>Track career opportunities</li>
          </ul>

          <div style="text-align: center; margin-bottom: 32px;">
            <a href="${env.frontendUrl || 'http://localhost:3000'}/login" style="display: inline-block; background-color: #2563EB; color: #ffffff; padding: 14px 28px; text-decoration: none; border-radius: 6px; font-weight: bold; font-size: 16px;">Log In to Your Account</a>
          </div>

          <div style="background-color: #27272A; padding: 16px; border-radius: 8px; font-size: 14px; color: #D4D4D8; line-height: 1.5;">
            <strong>Security Notice:</strong> For your protection, your password is never stored in plain text and will never be sent to you by email.
          </div>
        </div>
      `
    };

    await transporter.sendMail(mailOptions);
    return true;
  } catch (error) {
    console.error('[EmailService] Error sending welcome email:', error.message);
    return false;
  }
}

async function sendPasswordResetEmail(user, resetUrl) {
  if (!transporter) {
    console.log(`[EmailService] SMTP not set. Password reset link for ${user.email}: ${resetUrl}`);
    return false;
  }

  try {
    const mailOptions = {
      from: `"${env.mailFrom || 'Campus Cognition'}" <${env.mailUsername}>`,
      to: user.email,
      subject: 'Reset Your Campus Cognition Password',
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto; padding: 20px;">
          <h2 style="color: #1E3E62;">Password Reset Request</h2>
          <p>Hi ${user.username},</p>
          <p>We received a request to reset your password. Click the link below to set a new password:</p>
          <p style="margin: 25px 0;">
            <a href="${resetUrl}" style="background-color: #1E3E62; color: #ffffff; padding: 12px 24px; text-decoration: none; border-radius: 6px; font-weight: bold;">Reset Password</a>
          </p>
          <p>If you did not request this, please ignore this email.</p>
          <hr style="border: none; border-top: 1px solid #eee; margin: 20px 0;" />
          <p style="font-size: 12px; color: #888;">Campus Cognition Security</p>
        </div>
      `
    };

    await transporter.sendMail(mailOptions);
    return true;
  } catch (error) {
    console.error('[EmailService] Error sending password reset email:', error.message);
    return false;
  }
}

module.exports = {
  sendWelcomeEmail,
  sendPasswordResetEmail
};
